server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -233,12 +284,22 @@ public TemporalAccessor parse(String input) {
      */
     private TemporalAccessor doParse(String input) {
         if (parsers.size() > 1) {
+            Object object = null;
+            DateTimeFormatter lastParsedformatter = null;
             for (DateTimeFormatter formatter : parsers) {
                 ParsePosition pos = new ParsePosition(0);
-                Object object = formatter.toFormat().parseObject(input, pos);
+                object = formatter.toFormat().parseObject(input, pos);
                 if (parsingSucceeded(object, input, pos)) {
-                    return (TemporalAccessor) object;
+                    lastParsedformatter = formatter;
+                    break;
+                }
+            }
+            if (lastParsedformatter != null) {
+                if (canCacheLastParsedFormatter && lastParsedformatter != parsers.get(0)) {
+                    parsers.remove(lastParsedformatter);
```

Comment:
So here we are @CaptainDredge , apparently `JavaDateFormatter` violates thread safety guarantees and, well, we have `concurrent_modification_exception`:

```
     but: [<MapperParsingException[failed to parse field [date] of type [date] in document with id 'Pm2b-4oB5PIidmHQhuqx'. Preview of field's value: '2012-10-29T01:00:00.000+01:00']; nested: NotSerializableExceptionWrapper[concurrent_modification_exception: null];>]
```

Reply:
Yes, my understanding was that there's a single writer thread per shard. Maybe concurrent segment search changes that? Anyways I've now used concurrent friendly collection type `copyOnWriteArrayList` since its optimized for cases where reads >> writes which should be the case most of the times since on a specific shard the indexing workload remains same